extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SkipConcurrentExecutionInvoker.java

```diff
@@ -11,45 +13,35 @@
 import io.quarkus.scheduler.SkippedExecution;
 
 /**
- * A scheduled invoker wrapper that skips concurrent executions.
+ * An invoker wrapper that skips concurrent executions.
  *
  * @see Scheduled#concurrentExecution()
  * @see io.quarkus.scheduler.Scheduled.ConcurrentExecution#SKIP
  */
-public final class SkipConcurrentExecutionInvoker implements ScheduledInvoker {
+public final class SkipConcurrentExecutionInvoker extends DelegateInvoker {
 
-    private static final Logger LOGGER = Logger.getLogger(SkipConcurrentExecutionInvoker.class);
+    private static final Logger LOG = Logger.getLogger(SkipConcurrentExecutionInvoker.class);
 
     private final AtomicBoolean running;
-    private final ScheduledInvoker delegate;
     private final Event<SkippedExecution> event;
 
     public SkipConcurrentExecutionInvoker(ScheduledInvoker delegate, Event<SkippedExecution> event) {
+        super(delegate);
         this.running = new AtomicBoolean(false);
-        this.delegate = delegate;
         this.event = event;
     }
 
     @Override
-    public void invoke(ScheduledExecution execution) throws Exception {
+    public CompletionStage<Void> invoke(ScheduledExecution execution) throws Exception {
         if (running.compareAndSet(false, true)) {
-            try {
-                delegate.invoke(execution);
-            } finally {
-                running.set(false);
-            }
-        } else {
-            LOGGER.debugf("Skipped scheduled invoker execution: %s", delegate.getClass().getName());
-            SkippedExecution payload = new SkippedExecution(execution,
-                    "The scheduled method should not be executed concurrently");
-            event.fire(payload);
-            event.fireAsync(payload);
+            return delegate.invoke(execution).whenComplete((r, t) -> running.set(false));
         }
-    }
-
-    @Override
-    public void invokeBean(ScheduledExecution param) {
-        throw new UnsupportedOperationException();
+        LOG.debugf("Skipped scheduled invoker execution: %s", delegate.getClass().getName());
+        SkippedExecution payload = new SkippedExecution(execution,
+                "The scheduled method should not be executed concurrently");
+        event.fire(payload);
+        event.fireAsync(payload);
```

Comment:
I barely remember observers in CDI, but should we wait for the async one to complete before returning the completion stage?

Reply:
I don't think we should. It's more like a fire-and-forget style of communication...